tensorboard/plugins/graph/tf_graph_controls/tf-graph-controls.ts

```diff
@@ -12,397 +12,546 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-        // expand-more button points downwards, and the expand-less button points
-        // upwards. For most collapsibles, this works because the collapsibles
-        // expand in the downwards direction. This collapsible expands upwards
-        // though, so we reverse the icons.
-        return legendOpened ? "expand-more" : "expand-less";
-      },
-      _getSelectionOpGraphDisabled(datasets, _selectedRunIndex, _selectedTagIndex) {
-        return !datasets[_selectedRunIndex] ||
-            !datasets[_selectedRunIndex].tags[_selectedTagIndex] ||
-            !datasets[_selectedRunIndex].tags[_selectedTagIndex].opGraph;
-      },
-      _getSelectionProfileDisabled(datasets, _selectedRunIndex, _selectedTagIndex) {
-        return !datasets[_selectedRunIndex] ||
-            !datasets[_selectedRunIndex].tags[_selectedTagIndex] ||
-            !datasets[_selectedRunIndex].tags[_selectedTagIndex].profile;
-      },
-      _getSelectionConceptualGraphDisabled(datasets, _selectedRunIndex, _selectedTagIndex) {
-        return !datasets[_selectedRunIndex] ||
-            !datasets[_selectedRunIndex].tags[_selectedTagIndex] ||
-            !datasets[_selectedRunIndex].tags[_selectedTagIndex].conceptualGraph;
-      },
+      });
+      return {
+        device: device,
+        suffix: suffixes[i],
+        used: devicesForStats[device],
+        ignoredMsg: ignoredMsg,
+      };
     });
-})(); // Closing private scope.
\ No newline at end of file
+  },
+
+  _deviceCheckboxClicked: function(event: Event): void {
+    // Update the device map.
+    const input = (event.target as HTMLInputElement);
+    const devicesForStats: DeviceForStats = Object.assign(
+        {}, this.devicesForStats);
+    const device = input.value;
+    if (input.checked) {
+      devicesForStats[device] = true;
+    } else {
+      delete devicesForStats[device];
+    }
+    this.set('devicesForStats', devicesForStats);
+  },
+
+  _numTags: function(datasets: Dataset, _selectedRunIndex: number): number {
+    return this._getTags(datasets, _selectedRunIndex).length;
+  },
+
+  _getTags: function(
+      datasets: Dataset, _selectedRunIndex: number): TagItem[] {
+
+    if (!datasets || !datasets[_selectedRunIndex]) {
+      return [];
+    }
+    return datasets[_selectedRunIndex].tags;
+  },
+
+  fit: function(): void {
+    // TODO(stephanwlee): document -> this. A WebComponent should not use
+    // document.querySelector. tf-graph renders tf-graph-scene#scene
+    // which means this component has an incomplete API and encapsulation is
+    // leaking in a bad way.
+    // Constructor is not exported and is not typed.
+    const tfGraphScene: any = document.querySelector('#scene');
+    tfGraphScene.fit();
+  },
+
+  _isGradientColoring: function(
+      stats: tf.graph.proto.StepStats, colorBy: ColorBy): boolean {
+
+    return GRADIENT_COMPATIBLE_COLOR_BY.has(colorBy) &&
+        this._statsNotNull(stats);
```

Comment:
Is this not a functional change? (`stats != null` → `stats !== null`
through a function call)

Suggest removing this function call: IIUC, the function itself is only
needed due to Polymer insanity; there’s no reason to extract a function
for `(x) => x != null`.